design/one-pager-stack-relationship-labels.md

```diff
@@ -0,0 +1,166 @@
+# Stack Parent/Child Relationship Label Spec
+
+- Owner: Steven Rathbauer (@rathpc)
+- Reviewers: Crossplane Maintainers
+- Status: Draft
+
+## Proposal
+
+[crossplaneio/crossplane#934](https://github.com/crossplaneio/crossplane/issues/934)
+
+## Problem
+
+There is currently no easy way to get a list of all resources that are owned by a parent resource.
+We can leverage CRD labels to make this easier. As long as we have a standard format for doing so.
+
+Per the proposal, it seems that the closest example of what I am referring to is shown on 
+[this page](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels) 
+within the `app.kubernetes.io/part-of` label. However since we are not exactly using these labels 
```

Comment:
can you expand on how our goals differ from what's described on that page? I liked how that page was saying:

>  A common set of labels allows tools to work interoperably, describing objects in a common manner that all tools can understand.

I'd love to understand specifically what we're doing that is divergent from the goals on that page, can you add details here?

Reply:
I think @suskin might be able to convey the reasoning we decided to sway away from using those predefined labels better than I can.

Reply:
@suskin do you remember any of the details about why we wouldn't want to use the common upstream labels here?

Reply:
@jbw976 @rathpc the reason is summarized in this passage, which can be found in the introductory section:

> The metadata is organized around the concept of an application.

Essentially, because the labels are organized around the concept of an application, they do not really fit our concept of tracing multiple levels of parent/child relationships. Consider [the label descriptions](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels). They are missing the concept of a hierarchy like what we're trying to build. The label which comes the closest is `part-of`, which reads `The name of a higher level application this one is part of` (example: `wordpress`). None of the other labels are for describing a hierarchical relationship.

Let's make things more concrete with an example. We **could use** `part-of` and/or `instance` to point to parents if we wanted to. But this is not ideal. Consider the following example hierarchy, where `A -> B` means A is a parent of B:

```
WordpressInstance -> KubernetesCluster -> KubernetesApplication -> KubernetesApplicationResource
```

Semantically, the `app.kubernetes.io` labels "are organized around the concept of an application". Since the resources in the example hierarchy are all part of the same application (Wordpress), they should all have the same `instance` and `part-of` labels:

```
WordpressInstance labels
app.kubernetes.io/name: wordpress
app.kubernetes.io/instance: wordpress-instance-efafe342
app.kubernetes.io/component: application-claim
app.kubernetes.io/part-of: wordpress

KubernetesCluster labels
app.kubernetes.io/name: kubernetes
app.kubernetes.io/instance: wordpress-instance-efafe342
app.kubernetes.io/component: cluster
app.kubernetes.io/part-of: wordpress

KubernetesApplication labels
app.kubernetes.io/name: kubernetes-application
app.kubernetes.io/instance: wordpress-instance-efafe342
app.kubernetes.io/component: workload
app.kubernetes.io/part-of: wordpress

KubernetesApplicationResource labels
app.kubernetes.io/name: kubernetes-resource
app.kubernetes.io/instance: wordpress-instance-efafe342
app.kubernetes.io/component: workload-resource
app.kubernetes.io/part-of: wordpress
```

However, what we want is for each one to point back one step in the hierarchy. And that would contradict organizing the labels around an application. The `app.kubernetes.io` labels are still useful, just not for describing a multi-level hierarchy. What do you think? Did that make sense? If this question came up, maybe it would make sense to try to answer it in a little more detail in the document.

Reply:
Thank you @suskin! I have added most of this into the document to memorialize these findings and decisions